### ISSUE TITLE
Authentication cannot complete while a resolver is pending

### ISSUE THREAD

**Issue by jkeech** [microsoft/vscode#182217]
Does this issue occur when all extensions are disabled?: No

- VS Code Version: 1.78.1
- OS Version: macOS 13.3.1

Steps to Reproduce:

1. Create a codespace and connect to it
2. Disconnect from the codespace (`Remote: Close Remote Connection`)
3. Sign out in the account menu
4. File -> Open Recent -> select the codespace
5. A notification pops up asking to sign in: 
![image](https://github.com/microsoft/vscode/assets/746020/c4cf46cf-d78b-4790-b533-efe12316f157)
6. We would like to automatically sign in at this point instead of giving the notification, but we cannot today since the callback for authentication doesn't complete until the resolver finishes. The user has to perform the sign in in a different VS Code window or exit the resolver, sign in, and then reconnect again.

When we change our extension to start the auth flow while the resolver is pending, the user is able to go through the OAuth flow in the browser, but VS Code does not handle the callback while the resolver is pending. VS Code remains stuck waiting for sign in to finish.
![image](https://github.com/microsoft/vscode/assets/746020/e853ffa4-225a-4cbc-bee2-f33b3f315dd1)

`GitHub Authentication` output ends with the following:
```
2023-05-11 10:21:02.199 [info] Logging in for the following scopes: codespace read:user repo user:email
2023-05-11 10:21:02.200 [info] Trying without local server... (codespace read:user repo user:email)
```

Once the resolver process finally exits, the popup for the callback uri from a few minutes earlier appears:
![image](https://github.com/microsoft/vscode/assets/746020/46b379aa-0cd6-4e37-b8db-67f47f6d5951)

We would like to be able to handle the authentication callbacks while the resolver is still pending so that the user can complete the authentication flow and then finish connecting to the codespace after that. I'm happy to provide a local build of our extension that prompts for authentication automatically during the resolver instead of surfacing the notification if that helps with the repro/investigation. Thanks!

cc @alexdima @TylerLeonhardt @joyceerhl

**Comment by TylerLeonhardt** [comment]
Can you see if the GitHub Authentication is listed when you run the command: `Show Running Extensions`

**Comment by jkeech** [comment]
For reference, `Show Running Extensions` doesn't seem to work while the resolver is pending, but there is output in the `GitHub Authentication` output channel, so I'm guessing that the extension is running. It's able to kick off the auth flow just fine, just not able to complete the flow with the callback URI handling.

![image](https://github.com/microsoft/vscode/assets/746020/8078b8db-9666-4f31-80ed-67a669f2cb0d)

**Comment by TylerLeonhardt** [comment]
@jkeech if you hit Cancel, and follow the notifications do you eventually get unblocked? We have other auth flows to workaround certain issues. Just wanna narrow down the problem

**Comment by jkeech** [comment]
> @jkeech if you hit Cancel, and follow the notifications do you eventually get unblocked? We have other auth flows to workaround certain issues. Just wanna narrow down the problem

Yes! The local server mode succeeds during the resolver. And then once the resolver finishes, the original non-local-server flow's URI handling modal dialog pops up. Is there a way to force login to go through the local server mode from the beginning? If so, are there any downsides in using the local server mode?

**Comment by TylerLeonhardt** [comment]
I'd rather try to figure out how to get the non-local server flow working because that's the best and fastest flow, but we should be able to detect if the extension is running with a remote and use the localhost flow... however, I'd really like to only do that if the resolver is still resolving... not when it's finished resolving.

**Comment by joaomoreno** [comment]
Needs some understanding by breakpointing here: https://github.com/microsoft/vscode/blob/5860847717cd4ca232a2001557eedf9b97027229/src/vs/workbench/services/extensions/browser/extensionUrlHandler.ts#L140-L141

**Comment by alexdima** [comment]
@joaomoreno I speculate that execution holds up at https://github.com/microsoft/vscode/blob/5860847717cd4ca232a2001557eedf9b97027229/src/vs/workbench/services/extensions/browser/extensionUrlHandler.ts#L210

During the time when a resolver is running but before it returns, the `activateByEvent` call will be blocked by the resolver finishing. You can use `ActivationKind.Immediate` to force the call to proceed with activating the extension even during this special startup time.

**Comment by TylerLeonhardt** [comment]
This does seem to work fine with Tunnels. I suspect this is fixed, but @jkeech has to make a chance in the Codespaces extension to truly verify this so he's going to give that a go today.

**Comment by TylerLeonhardt** [comment]
@jkeech confirmed this is working now for him. Marking as verified!
